File: webflux-dialogue/src/test/java/com/study/webflux/rag/domain/monitoring/model/StageSnapshotTest.java
package com.study.webflux.rag.domain.monitoring.model;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class StageSnapshotTest {

	@Test
	@DisplayName("attributes는 null 이어도 빈 맵으로 정규화한다")
	void constructor_shouldNormalizeNullAttributesToEmptyMap() {
		StageSnapshot snapshot = new StageSnapshot(DialoguePipelineStage.RETRIEVAL,
			StageStatus.COMPLETED,
			Instant.now(),
			Instant.now(),
			10L,
			null);

		assertThat(snapshot.attributes()).isEmpty();
	}

	@Test
	@DisplayName("attributes는 외부 변경에 오염되지 않는 불변 맵으로 보관한다")
	void constructor_shouldDefensivelyCopyAttributes() {
		Map<String, Object> source = new LinkedHashMap<>();
		source.put("documentCount", 3);

		StageSnapshot snapshot = new StageSnapshot(DialoguePipelineStage.RETRIEVAL,
			StageStatus.COMPLETED,
			Instant.now(),
			Instant.now(),
			10L,
			source);

		source.put("documentCount", 9);
		source.put("memoryCount", 2);

		assertThat(snapshot.attributes())
			.containsEntry("documentCount", 3)
			.doesNotContainKey("memoryCount");
		assertThatThrownBy(() -> snapshot.attributes().put("another", 1))
			.isInstanceOf(UnsupportedOperationException.class);
	}
}

File: app-miyou/src/test/kotlin/com/miyou/app/domain/monitoring/model/StageSnapshotTest.kt
package com.miyou.app.domain.monitoring.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.Instant

class StageSnapshotTest {
    @Test
    @DisplayName("속성 맵이 비어 있어도 스냅샷을 생성할 수 있다")
    fun constructor_shouldAllowEmptyAttributes() {
        val snapshot =
            StageSnapshot(
                DialoguePipelineStage.RETRIEVAL,
                StageStatus.COMPLETED,
                Instant.now(),
                Instant.now(),
                10L,
                emptyMap(),
            )

        assertThat(snapshot.attributes).isEmpty()
    }

    @Test
    @DisplayName("전달한 속성 맵을 그대로 보존한다")
    fun constructor_shouldPreserveAttributes() {
        val snapshot =
            StageSnapshot(
                DialoguePipelineStage.RETRIEVAL,
                StageStatus.COMPLETED,
                Instant.now(),
                Instant.now(),
                10L,
                linkedMapOf("documentCount" to 3, "memoryCount" to 2),
            )

        assertThat(snapshot.attributes)
            .containsEntry("documentCount", 3)
            .containsEntry("memoryCount", 2)
    }
}
